File: python/detectionformats/amplitude.py
#!/usr/bin/env python

#stdlib imports
import json

class Amplitude:
    """ Amplitude - a conversion class used to create, parse, and validate  
        amplitude data as part of detection data.
    """
    # json keys
    AMPLITUDE_KEY = "Amplitude"
    PERIOD_KEY = "Period"
    SNR_KEY = "SNR"

    def __init__(self, newAmplitude=None, newPeriod=None, newSNR=None):
        """Initialize the amplitude object. Constructs an empty object
           if all arguments are None

        Args:
            newAmplitude: an optional Number containing the desired amplitude
                measurement as a float
            newPeriod: an optional Number containing the desired period
                measurement as a float
            newSNR: an optional Number containing the desired signal to noise
                measurement as a float
        Returns:
            Nothing
        Raises:
            Nothing
        """
        if newAmplitude is not None:
            self.amplitude = newAmplitude
        if newPeriod is not None:
            self.period = newPeriod
        if newSNR is not None:
            self.snr = newSNR

    def fromJSONString(self, jsonString):
        """Populates the object from a json formatted string

        Args:
            jsonString: a required String containing the json formatted text
        Returns:
            Nothing
        Raises:
            Nothing
        """
        jsonObject = json.loads(jsonString)
        self.fromDict(jsonObject)

    def fromDict(self, aDict):
        """Populates the object from a dictionary

        Args:
            aDict: a required dictionary
        Returns:
            Nothing
        Raises:
            Nothing
        """
        if self.AMPLITUDE_KEY in aDict:
            self.amplitude = aDict[self.AMPLITUDE_KEY]
        if self.PERIOD_KEY in aDict:
            self.period = aDict[self.PERIOD_KEY]
        if self.SNR_KEY in aDict:
            self.snr = aDict[self.SNR_KEY]

    def toJSONString(self):
        """Converts the object to a json formatted string

        Args:
            None
        Returns:
            The JSON formatted message as a String
        Raises:
            Nothing
        """
        jsonObject = self.toDict()

        return json.dumps(jsonObject, ensure_ascii=False)

    def toDict(self):
        """Converts the object to a dictionary

        Args:
            None
        Returns:
            The dictionary
        Raises:
            Nothing
        """
        aDict = {}
        if hasattr(self, 'amplitude'):
            aDict[self.AMPLITUDE_KEY] = self.amplitude

        if hasattr(self, 'period'):
            aDict[self.PERIOD_KEY] = self.period

        if hasattr(self, 'snr'):
            aDict[self.SNR_KEY] = self.snr

        return aDict

    def isValid(self):
        """Checks to see if the object is valid

        Args:
            None
        Returns:
            True if the object is valid, False otherwise
        Raises:
            Nothing
        """
        errorList = self.getErrors()

        if len(errorList) == 0:
            return True
        else:
            return False

    def getErrors(self):
        """Gets a list of object validation errors

        Args:
            None
        Returns:
            A List of Strings containing the validation error messages
        Raises:
            Nothing
        """
        errorList = []

        # all members are optional

        return errorList

    def isEmpty(self):
        """Checks to see if the object is empty

        Args:
            None
        Returns:
            True if the object is empty, False otherwise
        Raises:
            Nothing
        """
        if hasattr(self, 'amplitude'):
            return False

        if hasattr(self, 'period'):
            return False

        if hasattr(self, 'snr'):
            return False

        return True

File: java/src/main/java/gov/usgs/detectionformats/Amplitude.java
package gov.usgs.detectionformats;

import java.util.ArrayList;
import org.json.simple.JSONObject;

/**
 * a conversion class used to create, parse, and validate amplitude data as part of detection data.
 *
 * @author U.S. Geological Survey &lt;jpatton at usgs.gov&gt;
 */
public class Amplitude implements DetectionInt {

  /** JSON Keys */
  public static final String AMPLITUDE_KEY = "Amplitude";

  public static final String PERIOD_KEY = "Period";
  public static final String SNR_KEY = "SNR";

  /** Optional double containing the Amplitude */
  private final Double amplitude;

  /** Optional double containing the Period */
  private final Double period;

  /** Optional double containing the SNR */
  private final Double snr;

  /** The constructor for the Amplitude class. Initializes members to null values. */
  public Amplitude() {

    amplitude = null;
    period = null;
    snr = null;
  }

  /**
   * The advanced constructor for the Amplitude class. Initializes members to provided values.
   *
   * @param newAmplitude - A double containing the Amplitude to use (null omit)
   * @param newPeriod - A double containing the Period to use (null omit)
   * @param newSNR - A double containing the SNR to use (null omit)
   */
  public Amplitude(final Double newAmplitude, final Double newPeriod, final Double newSNR) {

    amplitude = newAmplitude;
    period = newPeriod;
    snr = newSNR;
  }

  /**
   * Constructs the class from a JSONObject, populating members
   *
   * @param newJSONObject - A JSONObject.
   */
  public Amplitude(JSONObject newJSONObject) {

    // optional values
    // amplitude
    if (newJSONObject.containsKey(AMPLITUDE_KEY)) {
      amplitude = Double.valueOf(newJSONObject.get(AMPLITUDE_KEY).toString());
    } else {
      amplitude = null;
    }

    // period
    if (newJSONObject.containsKey(PERIOD_KEY)) {
      period = Double.valueOf(newJSONObject.get(PERIOD_KEY).toString());
    } else {
      period = null;
    }

    // snr
    if (newJSONObject.containsKey(SNR_KEY)) {
      snr = Double.valueOf(newJSONObject.get(SNR_KEY).toString());
    } else {
      snr = null;
    }
  }

  /**
   * Converts the contents of the class to a JSON object
   *
   * @return Returns a JSONObject containing the class contents
   */
  @SuppressWarnings("unchecked")
  public JSONObject toJSON() {

    JSONObject newJSONObject = new JSONObject();

    Double jsonAmplitude = getAmplitude();
    Double jsonPeriod = getPeriod();
    Double jsonSnr = getSNR();

    // optional values
    // amplitude
    if (jsonAmplitude != null) {
      newJSONObject.put(AMPLITUDE_KEY, jsonAmplitude);
    }

    // period
    if (jsonPeriod != null) {
      newJSONObject.put(PERIOD_KEY, jsonPeriod);
    }

    // snr
    if (jsonSnr != null) {
      newJSONObject.put(SNR_KEY, jsonSnr);
    }

    // return valid object
    return (newJSONObject);
  }

  /**
   * Validates the class.
   *
   * @return Returns true if successful
   */
  public boolean isValid() {
    if (getErrors().size() == 0) {
      return (true);
    } else {
      return (false);
    }
  }

  /**
   * Gets any validation errors in the class.
   *
   * @return Returns a List&lt;String&gt; of any errors found
   */
  public ArrayList<String> getErrors() {
    Double jsonPeriod = getPeriod();
    Double jsonSNR = getSNR();

    ArrayList<String> errorList = new ArrayList<String>();

    // optional data
    // period
    if (jsonPeriod != null) {
      if (jsonPeriod < 0) {
        errorList.add("Invalid Period in Amplitude class.");
      }
    }

    // SNR
    if (jsonSNR != null) {
      if ((jsonSNR < 0) || (jsonSNR > 9999999999.999)) {
        errorList.add(
            "SNR in Amplitude Class is not in the valid " + "range of 0 to 9999999999.999");
      }
    }

    // success
    return (errorList);
  }

  /**
   * Checks to see if this object is empty
   *
   * @return Returns true if empty, false otherwise.
   */
  public boolean isEmpty() {
    if (getAmplitude() != null) {
      return (false);
    }
    if (getPeriod() != null) {
      return (false);
    }
    if (getSNR() != null) {
      return (false);
    }

    return (true);
  }

  /** @return the amplitude */
  public Double getAmplitude() {
    return amplitude;
  }

  /** @return the period */
  public Double getPeriod() {
    return period;
  }

  /** @return the snr */
  public Double getSNR() {
    return snr;
  }
}
